### Detailed Analysis of Individual Contributions to Presidential, Senate, and House Committees


**Introduction:** In this project, we conduct an analysis of individual contributions to Federal Committees such as presidential, senate, and house committees for the period from June 20, 2024, to July 24, 2024. This analysis will involve data collection, data processing, and data analysis using Apache Spark. The goal is to gain insights into the patterns and trends of individual contributions to political committees during this period.

**Project Submission by Prajwal Kusha**

### Data Collection (10%)

#### Step 1: Uploading the Data and unzipping it: June and July Data

Given that the FEC data downloaded from the FEC website is huge, approximately **3.37 GB**. We zip the required files and upload it on the instance to proceed with the next steps. 
* The data to be uploaded are named according to the month in the ZIP file downloaded from the FEC website.
* We proceed uploading the following data as available: **itcont_2024_20240620_20240709.txt, itcont_2024_20240710_20240723.txt** and the header file named **indiv_header_file.csv**

**Code Explanation:**

* `!unzip` A shell command to unzip (extract) a compressed .zip file.
* `-o`: Overwrites existing files in the destination directory without prompting for confirmation.
* `itcont_2024_20240620_20240709.txt.zip`: The input file to be unzipped, which contains data for June contributions.
* `-d ./Fec_Jun`: Specifies the destination directory (./Fec_Jun) where the unzipped files will be extracted. If the directory does not exist, it will be created automatically.

In [1]:
!unzip -o itcont_2024_20240620_20240709.txt.zip -d ./Fec_Jun
!unzip -o itcont_2024_20240710_20240723.txt.zip -d ./Fec_Jul

Archive:  itcont_2024_20240620_20240709.txt.zip
  inflating: ./Fec_Jun/itcont_2024_20240620_20240709.txt  
  inflating: ./Fec_Jun/__MACOSX/._itcont_2024_20240620_20240709.txt  
Archive:  itcont_2024_20240710_20240723.txt.zip
  inflating: ./Fec_Jul/itcont_2024_20240710_20240723.txt  
  inflating: ./Fec_Jul/__MACOSX/._itcont_2024_20240710_20240723.txt  


#### Step 2: Removing the zip files: 
We do this so that we can save space on the instance. 

**Code Explanation:**
* `!rm` - removes the files mentioned

In [2]:
!rm itcont_2024_20240620_20240709.txt.zip
!rm itcont_2024_20240710_20240723.txt.zip

### Data Processing

#### Step 3: Converting the unzipped .txt file to .csv - June, July Data

**Code Explanation:**

* `!csvformat` Is a utility from the csvkit package used for manipulating and reformatting CSV files
* `-d`Specifies the delimiter used in the input file.
* `"|"`Indicates that the input file is pipe-delimited (| separates each field in the file)
* `>` Sends the output of the csvformat command to a new file rather than displaying it in the terminal.

In [3]:
!csvformat -d "|" Fec_Jun/itcont_2024_20240620_20240709.txt > Fec_Jun/June24.csv

In [4]:
!csvformat -d "|" Fec_Jul/itcont_2024_20240710_20240723.txt > Fec_Jul/July24.csv

#### Step 4: Word count of the actual (.txt) file and the converted (.csv) file

**Code Explanation:**

* `wc` Word Count is usually used to count lines, characters and words in a file. 
* `-l` Specifies that the command should count only the number of lines in the file.

**Inference: We see that the conversion has been successful. Both the files have the same number of lines**
* June24 = 1944602
* July24 = 1902192

In [5]:
!wc -l Fec_Jun/itcont_2024_20240620_20240709.txt
!wc -l Fec_Jun/June24.csv

1944602 Fec_Jun/itcont_2024_20240620_20240709.txt
1944602 Fec_Jun/June24.csv


In [6]:
!wc -l Fec_Jul/itcont_2024_20240710_20240723.txt
!wc -l Fec_Jul/July24.csv

1902192 Fec_Jul/itcont_2024_20240710_20240723.txt
1902192 Fec_Jul/July24.csv


#### Step 5: Combine the header and data files using concatenate

**Code Explanation:**

* `!cat` The cat command is used combine multiple files into a single file.
* The cat takes the header file and stacks on top of the June24 and July24 files and combines all 3 to provide the JunJul_24 file. 

In [7]:
!cat indiv_header_file.csv Fec_Jun/June24.csv Fec_Jul/July24.csv > JunJul_24.csv

#### Step 6: Run word count to check if the combining was successful. 

* `!wc -l` Is used to count number of lines. 
* We can observe that 1944602 + 1902192 + 1 = **3846795 = JunJul_24** (Hence combining was successful)

In [8]:
!wc -l JunJul_24.csv

3846795 JunJul_24.csv


#### Step 7: Understanding the Data - Print header column of the data

**Code Explanation:**

* `!csvcut` The command displays the column names in a file
* `-n` Lists all column headers (names) in the CSV file along with their indices.

In [9]:
!csvcut -n JunJul_24.csv

  1: CMTE_ID
  2: AMNDT_IND
  3: RPT_TP
  4: TRANSACTION_PGI
  5: IMAGE_NUM
  6: TRANSACTION_TP
  7: ENTITY_TP
  8: NAME
  9: CITY
 10: STATE
 11: ZIP_CODE
 12: EMPLOYER
 13: OCCUPATION
 14: TRANSACTION_DT
 15: TRANSACTION_AMT
 16: OTHER_ID
 17: TRAN_ID
 18: FILE_NUM
 19: MEMO_CD
 20: MEMO_TEXT
 21: SUB_ID


#### Step 8: Understanding the data - Print top 10 rows of the data. 

**Code Explanation:**

* `!head`is a shell command used to display the beginning (first few lines) of a file.
* `-n 10`Specifies the number of lines to display

In [10]:
!head -n 10 JunJul_24.csv 

CMTE_ID,AMNDT_IND,RPT_TP,TRANSACTION_PGI,IMAGE_NUM,TRANSACTION_TP,ENTITY_TP,NAME,CITY,STATE,ZIP_CODE,EMPLOYER,OCCUPATION,TRANSACTION_DT,TRANSACTION_AMT,OTHER_ID,TRAN_ID,FILE_NUM,MEMO_CD,MEMO_TEXT,SUB_ID
C00849273,T,TER,P2024,202407159660924932,22Y,IND,"REPASS, JAMES",HILLSBORO,VA,20132,,,06202024,400,,SB28A.5248,1802122,,,4071620241974571456
C00828541,A,M7,P2024,202408159666207641,15,IND,"MANSFIELD, RISA",COLUMBUS,MS,39705,RETIRED,RETIRED,06202024,125,,SA17A.121694463,1810271,,,4081620242013476924
C00828541,A,M7,P2024,202408159666207641,15,IND,"MANTZOURANIS, VASILIKE",KENSINGTON,MD,20895,RETIRED,RETIRED,06202024,150,,SA17A.121693527,1810271,,,4081620242013476925
C00828541,A,M7,P2024,202408159666209186,15,IND,"SCOTT, HORACE EDWARD",HANCEVILLE,AL,35077,HORACE F SEPTT,RETAIR,06202024,100,,SA17A.121694767,1810271,,,4081620242013481558
C00828541,A,M7,P2024,202408159666204850,15,IND,"BRUSA, S E",LODI,CA,95242,HOMEMAKER,HOMEMAKER,06202024,100,,SA17A.121694255,1810271,,,4081620242013468550
C00

#### Step 9: Initializing Spark Environment 

**Code Explanation**
* `import findspark` A Python library used to locate and initialize Apache Spark in your environment.
* `findspark.init()` Initializes Spark in the current Python environment.
* `from pyspark import SparkContext` The primary entry point for using PySpark's RDD (Resilient Distributed Dataset) API.
* `spark = SparkContext(appName='20241029')` Assigns a name to your Spark application, which will appear in the Spark UI or logs for tracking purposes.

In [11]:
import findspark
findspark.init()
from pyspark import SparkContext
spark = SparkContext(appName='20241029')
spark

Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
24/11/25 23:01:12 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


<SparkContext master=local[*] appName=20241029>

#### Step 10: Loading file onto Spark RDD 

* `FEC` The variable FEC stores the resulting RDD after loading the file
* `spark.textFile` A method provided by the SparkContext (spark) to read text files.

In [12]:
FEC = spark.textFile('JunJul_24.csv')

#### Step 11: Run word count to see if data load was successful

* `%time` Provides the time in which the task was completed
* `FEC.count()` Counts the number of lines from the data loaded on spark.

In [13]:
%time FEC.count()

[Stage 0:======================================================>  (21 + 1) / 22]

CPU times: user 17.9 ms, sys: 5.1 ms, total: 23 ms
Wall time: 5.82 s


3846795

### Data Analysis Part 1: RDD Processing (30%)

#### 1. The top 10 states in terms of the total count of contributions.

**Code Explanation**

* `from operator import add` The add function is used to sum values.
* `header = FEC.first()` Retrieves the first row of the RDD, which is assumed to be the header of the CSV file.
* `FEC.filter(lambda row: row != header)` Filters out the header row from the RDD, ensuring only data rows are processed.
* `row.replace('"', '')` Removes any double quotes (") from the row to avoid issues with splitting or processing.
* `row.split(',')` Splits each line of the CSV into a list of columns based on the comma (,) delimiter.
* `cols[10]` Refers to the column corresponding to STATE. Ensures the STATE column is not empty ('') or None.
* `.reduceByKey(add)` Aggregates the values for each state (key) by summing them using the add function.
* `-pair[1])` Orders the states by the total count of contributions (pair[1]) in descending order (-pair[1]).

**Note:** We know that Pyspark follows zero-based indexing for column positions in DataFrames and RDDs. This means that the first column is indexed as 0, the second column as 1, and so on. So ideally our state column should be column number 9 from our data. But in our code we have used column 10 to get the output, this is because the column 'NAME' has ("") which is wrongly predicted by pyspark and is hence considering the ("") to be a seperate column all together. Hence the columns after 'NAME' have an increment of 1 to them. Making column 'STATE' col[10] instead of 9. 

In [14]:
from operator import add
header = FEC.first()
top10 = FEC.filter(lambda row: row != header) \
    .map(lambda row: row.replace('"', '')) \
    .map(lambda row: row.split(',')) \
    .filter(lambda cols: cols[10] is not None and cols[10] != '') \
    .map(lambda cols: (cols[10], 1)) \
    .reduceByKey(add) \
    .takeOrdered(10, key=lambda pair: -pair[1])

# Print the results
print("Top 10 States by Total Count of Contributions:")
for state, count in top10:
    print(f"{state}\t{count}")

[Stage 2:===============>                                          (6 + 2) / 22]24/11/25 23:01:25 WARN GarbageCollectionMetrics: To enable non-built-in garbage collector(s) List(G1 Concurrent GC), users should configure it(them) to spark.eventLog.gcMetrics.youngGenerationGarbageCollectors or spark.eventLog.gcMetrics.oldGenerationGarbageCollectors
[Stage 3:======================================================>  (21 + 1) / 22]

Top 10 States by Total Count of Contributions:
CA	564941
TX	324187
FL	279517
NY	225839
WA	140701
PA	130224
VA	125452
IL	124756
AZ	114972
OH	109487


#### 2. The top 10 states in terms of the total (i.e., sum) amount of contributions

**Code Explanation**

* `cols: cols[10] is not None and cols[10] != ''` cols[10] (state column) is not empty or None.
* `cols[15] is not None and cols[15] != ''` cols[15] (transaction amount) is not empty or None.
* `replace('.', '', 1)` Removes the decimal point for validation.
* `isdigit()` Checks if the resulting string is numeric.
* `float(cols[15]` The contribution amount from cols[15], converted to a float.
* `:,.2f` The total contribution amount formatted with commas and two decimal places

**NOTE:** As mentioned earlier, every column after 'NAME' is incremented with 1. Hence 'TRANSACTION_AMT' is col[15]. 

In [15]:
from operator import add

header = FEC.first()

top10_sum = (
    FEC.filter(lambda row: row != header)  
    .map(lambda row: row.replace('"', '')) 
    .map(lambda row: row.split(','))  
    .filter(lambda cols: cols[10] is not None and cols[10] != '' and cols[15] is not None and cols[15] != '')  
    .filter(lambda cols: cols[15].replace('.', '', 1).isdigit())  
    .map(lambda cols: (cols[10], float(cols[15])))  
    .reduceByKey(add)  
    .takeOrdered(10, key=lambda pair: -pair[1])
)

# Print the results
print("\nTop 10 states by sum of contributions:")
for state, total in top10_sum:
    print(f"{state}\t{total:,.2f}")

[Stage 6:==============================================>          (18 + 2) / 22]


Top 10 states by sum of contributions:
CA	64,848,952,933.00
TX	55,447,769,793.00
IL	34,138,523,773.00
NY	30,559,012,259.00
FL	26,209,564,812.00
GA	23,380,003,290.00
VA	21,875,285,070.00
PA	21,375,962,094.00
NJ	20,531,652,271.00
MA	18,482,278,923.00


### Part 2: Data Frame API (30%)

#### Step 12: Proceed with initializing SQL on spark

In [16]:
from pyspark import SQLContext
sqlc = SQLContext(spark)
sqlc

/usr/local/lib/spark/python/pyspark/sql/context.py:113: FutureWarning: Deprecated in 3.0.0. Use SparkSession.builder.getOrCreate() instead.
  warnings.warn(


#### Step 13: Loading file on to the Dataframe

**Code Explanation**
* `sqlc.read.csv` Reads the file JunJul_24.csv into a Spark DataFrame.
* `header=True` Indicates that the first row of the CSV file contains column names (headers).
* `inferSchema=False` With inferSchema=False, all columns are read as strings, regardless of their actual data type.

**Why use inferSchema=False**
* **Performance Optimization:** Having inferSchema false is faster especially for large datasets, because it avoids the overhead of reading a portion of the data for type inference.
* **Consistency in Data Types:** A column might contain mostly numeric data but include some non-numeric values (e.g., "N/A" or "NULL"). If inferSchema=True, Spark may fail to parse such values or throw errors during operations.
* Given that our data contains null values, we choose to call inferSchema=False, as we can explicitly define the schema later using the schema parameter or by casting columns manually.

**NOTE:** From the data available, we see that **Transaction_DT** is in the format of **'06202024'**, but when **inferSchema=True** the data is interpreted as **'6202024'**. This is the major reason for us to call inferSchema=False, as we can **define the datatype manually** to provide the **most accurate output possible.**

In [17]:
FEC = sqlc.read.csv("JunJul_24.csv", header=True, inferSchema=False)

#### Step 14: Count the number of lines after loading data onto the Dataframe

* We can see that this matches with the number of lines in the actual file (JunJul_24.csv), hence data loading was successful

In [18]:
FEC.count()

3846794

#### Step 15: Print first 5 rows of data. 

* Run this to identify any errors if occured

In [19]:
FEC.take(5)

[Row(CMTE_ID='C00849273', AMNDT_IND='T', RPT_TP='TER', TRANSACTION_PGI='P2024', IMAGE_NUM='202407159660924932', TRANSACTION_TP='22Y', ENTITY_TP='IND', NAME='REPASS, JAMES', CITY='HILLSBORO', STATE='VA', ZIP_CODE='20132', EMPLOYER=None, OCCUPATION=None, TRANSACTION_DT='06202024', TRANSACTION_AMT='400', OTHER_ID=None, TRAN_ID='SB28A.5248', FILE_NUM='1802122', MEMO_CD=None, MEMO_TEXT=None, SUB_ID='4071620241974571456'),
 Row(CMTE_ID='C00828541', AMNDT_IND='A', RPT_TP='M7', TRANSACTION_PGI='P2024', IMAGE_NUM='202408159666207641', TRANSACTION_TP='15', ENTITY_TP='IND', NAME='MANSFIELD, RISA', CITY='COLUMBUS', STATE='MS', ZIP_CODE='39705', EMPLOYER='RETIRED', OCCUPATION='RETIRED', TRANSACTION_DT='06202024', TRANSACTION_AMT='125', OTHER_ID=None, TRAN_ID='SA17A.121694463', FILE_NUM='1810271', MEMO_CD=None, MEMO_TEXT=None, SUB_ID='4081620242013476924'),
 Row(CMTE_ID='C00828541', AMNDT_IND='A', RPT_TP='M7', TRANSACTION_PGI='P2024', IMAGE_NUM='202408159666207641', TRANSACTION_TP='15', ENTITY_TP='I

#### Step 16: Print Schema of the data
* We do this to understand the schema of the data that was loaded. 
* By default it has been assumed to be string because we have called inferSchema=False

In [20]:
FEC.printSchema()

root
 |-- CMTE_ID: string (nullable = true)
 |-- AMNDT_IND: string (nullable = true)
 |-- RPT_TP: string (nullable = true)
 |-- TRANSACTION_PGI: string (nullable = true)
 |-- IMAGE_NUM: string (nullable = true)
 |-- TRANSACTION_TP: string (nullable = true)
 |-- ENTITY_TP: string (nullable = true)
 |-- NAME: string (nullable = true)
 |-- CITY: string (nullable = true)
 |-- STATE: string (nullable = true)
 |-- ZIP_CODE: string (nullable = true)
 |-- EMPLOYER: string (nullable = true)
 |-- OCCUPATION: string (nullable = true)
 |-- TRANSACTION_DT: string (nullable = true)
 |-- TRANSACTION_AMT: string (nullable = true)
 |-- OTHER_ID: string (nullable = true)
 |-- TRAN_ID: string (nullable = true)
 |-- FILE_NUM: string (nullable = true)
 |-- MEMO_CD: string (nullable = true)
 |-- MEMO_TEXT: string (nullable = true)
 |-- SUB_ID: string (nullable = true)



#### Step 17: Importing functions and defining Schema for the data

* `pyspark.sql.functions` Contains useful functions for manipulating DataFrame columns
* `DateType`: Represents date values.
* `FloatType`: Represents floating-point numbers.
* `IntegerType`: Represents 32-bit signed integers.
* `StringType`: Represents string values.

**As mentioned in Step 13, we call inferSchema=False. Hence we define the datatype of Transaction_Amt below. 
* `cast(LongType())` Converts the column's data type to LongType (a long integer).
* This step is required to get the most accurate output. 

In [21]:
from pyspark.sql.functions import to_date, col, desc, format_number
from pyspark.sql.types import DateType, FloatType, IntegerType, StringType, LongType

FEC = sqlc.read.csv("JunJul_24.csv", header=True, inferSchema=False)

# Convert transaction amount
FEC = FEC.withColumn("TRANSACTION_AMT", 
            col("TRANSACTION_AMT").cast(LongType()))

#### Step 18: Part 2 -> Question 1

1. The contributor’s name, the committee ID the contributor contributed to, and the transaction amount for all contributions above $5000. Order the records by the transaction amount in descending order. Only show the top 10 rows.

**Code Explanation**
* `filter("TRANSACTION_AMT > 5000")` Only rows satisfying the condition (TRANSACTION_AMT > 5000) are retained.
* `select` Reduces the DataFrame to only the necessary columns for the analysis, improving efficiency and readability
* `desc("TRANSACTION_AMT")` Sorts the TRANSACTION_AMT column in descending order.
* `withColumn` Modifies the existing TRANSACTION_AMT column by applying formatting.
* `format_number("TRANSACTION_AMT", 2)` Formats the numeric values in the TRANSACTION_AMT column to include commas and two decimal places.
* `show(10)` Displays the first 10 rows of the processed DataFrame.
* `False` Prevents truncation of column values, ensuring the full content is displayed.

In [22]:
FEC.filter("TRANSACTION_AMT > 5000") \
   .select(FEC['NAME'], FEC['CMTE_ID'], FEC['TRANSACTION_AMT']) \
   .orderBy(desc("TRANSACTION_AMT")) \
   .withColumn("TRANSACTION_AMT", format_number("TRANSACTION_AMT", 2)) \
   .show(10, False)

[Stage 13:================================================>         (5 + 1) / 6]

+---------------------------------+---------+---------------+
|NAME                             |CMTE_ID  |TRANSACTION_AMT|
+---------------------------------+---------+---------------+
|MELLON, TIMOTHY                  |C00825851|50,000,000.00  |
|ONE NATION                       |C00571703|18,400,000.00  |
|SECURING AMERICAN GREATNESS      |C00881805|15,000,000.00  |
|FUTURE FORWARD USA ACTION        |C00669259|15,000,000.00  |
|CLEMENT, CHRISTINA, HH LOREN REV.|C00857128|12,000,000.00  |
|BLOOMBERG, MICHAEL R.            |C00495028|10,000,000.00  |
|SINGER, PAUL E.                  |C00504530|10,000,000.00  |
|SINGER, PAUL ELLIOTT             |C00571703|10,000,000.00  |
|BRICK BY BRICK FOUNDATION        |C00631549|5,000,000.00   |
|AMERICAN ACTION NETWORK          |C00504530|5,000,000.00   |
+---------------------------------+---------+---------------+
only showing top 10 rows



#### Step 19: Part 2 -> Question 2

2. Find the name of the individual and their total contributions. Order the records by the total transaction amount in descending order. Only show the top 10 rows.

**Code Explanation**
* `groupBy("NAME")` Groups rows in the DataFrame by the NAME column (contributor's name).
* `agg()` Specifies the aggregation operation to be performed on each group created by groupBy.
* `sum_col("TRANSACTION_AMT")` Sums the values in the TRANSACTION_AMT column for each contributor (NAME).
* `.alias("TOTAL_CONTRIBUTION")` Renames the aggregated column to TOTAL_CONTRIBUTION.
* `orderBy(desc("TOTAL_CONTRIBUTION"))` Sorts the DataFrame in descending order of the TOTAL_CONTRIBUTION column.  
* `format_number("TOTAL_CONTRIBUTION", 2)` Adds commas as thousand separators, for better view. 

In [23]:
from pyspark.sql.functions import sum as sum_col, desc, format_number

FEC.groupBy("NAME") \
   .agg(sum_col("TRANSACTION_AMT").alias("TOTAL_CONTRIBUTION")) \
   .orderBy(desc("TOTAL_CONTRIBUTION")) \
   .withColumn("TOTAL_CONTRIBUTION", format_number("TOTAL_CONTRIBUTION", 2)) \
   .show(10, False)

[Stage 16:>                                                         (0 + 2) / 2]

+---------------------------------+------------------+
|NAME                             |TOTAL_CONTRIBUTION|
+---------------------------------+------------------+
|MELLON, TIMOTHY                  |55,003,300.00     |
|ONE NATION                       |18,400,000.00     |
|FUTURE FORWARD USA ACTION        |15,190,058.00     |
|SECURING AMERICAN GREATNESS      |15,000,000.00     |
|MUSK, ELON                       |14,950,000.00     |
|CLEMENT, CHRISTINA, HH LOREN REV.|12,000,564.00     |
|SINGER, PAUL E.                  |10,150,000.00     |
|BLOOMBERG, MICHAEL R.            |10,014,900.00     |
|SINGER, PAUL ELLIOTT             |10,000,000.00     |
|AMERICAN ACTION NETWORK          |7,500,000.00      |
+---------------------------------+------------------+
only showing top 10 rows



### Part 3: SQL with Data Frame (30%)

#### Step 20: Run SQL queries on the data using PySpark's SQL engine.

* `createOrReplaceTempView` Creates a temporary SQL view named fec_data from the DataFrame FEC.

In [24]:
FEC.createOrReplaceTempView("fec_data")

#### Step 21: Run line count

* We run this to ensure the data load is successful and there are no rows missing. 

In [25]:
sqlc.sql("SELECT COUNT(*) FROM fec_data").show()

[Stage 17:================================================>         (5 + 1) / 6]

+--------+
|count(1)|
+--------+
| 3846794|
+--------+



#### Step 22: Part 3 -> Question 1
* Find the individual contributor’s name, the contributor's occupation, and the transaction amount. Order the records by the transaction amount in descending order. Only show the top 10 rows.

**Code Explanation**
* `sqlc.sql` is used to call sql to perform analysis 
* `SELECT` Specifies the columns to retrieve from the view:
* `NAME` The contributor's name.
* `OCCUPATION` The contributor's occupation.
* `TRANSACTION_AMT` The transaction amount of the contribution.
* `FROM fec_data` Indicates the data source (the fec_data temporary view).
* `ORDER BY TRANSACTION_AMT DESC` Sorts the rows by the TRANSACTION_AMT column in descending order.
* `LIMIT 10` Restricts the output to the top 10 rows after sorting.
* `.show()` Displays the results of the query in the console.
* `n=10` Displays up to 10 rows.
* `truncate=False` Ensures full column values are displayed without truncation.

In [26]:
sqlc.sql("""
    SELECT NAME, OCCUPATION, TRANSACTION_AMT
    FROM fec_data
    ORDER BY TRANSACTION_AMT DESC
    LIMIT 10
""").show(n=10, truncate=False)

[Stage 20:================================================>         (5 + 1) / 6]

+---------------------------------+-------------+---------------+
|NAME                             |OCCUPATION   |TRANSACTION_AMT|
+---------------------------------+-------------+---------------+
|MELLON, TIMOTHY                  |INVESTMENTS  |50000000       |
|ONE NATION                       |NULL         |18400000       |
|SECURING AMERICAN GREATNESS      |NULL         |15000000       |
|FUTURE FORWARD USA ACTION        |NULL         |15000000       |
|CLEMENT, CHRISTINA, HH LOREN REV.|SELF EMPLOYED|12000000       |
|BLOOMBERG, MICHAEL R.            |FOUNDER      |10000000       |
|SINGER, PAUL E.                  |PRESIDENT    |10000000       |
|SINGER, PAUL ELLIOTT             |PRESIDENT    |10000000       |
|BRICK BY BRICK FOUNDATION        |NULL         |5000000        |
|AMERICAN ACTION NETWORK          |NULL         |5000000        |
+---------------------------------+-------------+---------------+



#### Step 23: Part 3 -> Question 2
* The top 10 committees based on their total individual contributions on a given date. Provide the committee ID, the transaction date, and the transaction amount. Order the output by the total transaction amount in descending order. 

**Code Explanation**
* `SELECT` Specifies the columns to retrieve and the aggregation
* `AS total_transaction_amount` The result is aliased as total_transaction_amount.
* `FROM fec_data` Indicates the data source (the fec_data temporary view).
* `GROUP BY CMTE_ID, TRANSACTION_DT` Groups the rows by CMTE_ID and TRANSACTION_DT
* `ORDER BY total_transaction_amount DESC` Sorts the rows by the total_transaction_amount column in descending order.
* `LIMIT 10` Restricts the output to the top 10 rows after sorting.
* `.show()` Displays the results of the query in the console.
* `n=10` Displays up to 10 rows.
* `truncate=False` Ensures full column values are displayed without truncation.

**NOTE:** As we can see the output in TRANSACTION_DT is in the original format as our data, hence we call inferSchema=False in the earlier steps.

In [27]:
sqlc.sql("""
SELECT CMTE_ID, TRANSACTION_DT, SUM(TRANSACTION_AMT) AS total_transaction_amount
FROM fec_data
GROUP BY CMTE_ID, TRANSACTION_DT
ORDER BY total_transaction_amount DESC
LIMIT 10
""").show(n=10, truncate=False)

[Stage 21:================================================>         (5 + 1) / 6]

+---------+--------------+------------------------+
|CMTE_ID  |TRANSACTION_DT|total_transaction_amount|
+---------+--------------+------------------------+
|C00825851|07152024      |50000000                |
|C00401224|07212024      |26894750                |
|C00401224|07222024      |26418477                |
|C00571703|06282024      |18650000                |
|C00401224|07232024      |16219539                |
|C00669259|06262024      |16000000                |
|C00703975|07212024      |15996549                |
|C00881805|07182024      |15000000                |
|C00744946|07222024      |14057883                |
|C00825851|06262024      |14000000                |
+---------+--------------+------------------------+



#### Conclusion: Key Insights from Political Contributions Analysis
* The dataset reveals significant insights into political contributions, highlighting patterns in high-value donors, PACs - Political Action Committees, and geographic trends.

**Top States for Contributions:**
* **California (CA), Texas (TX), and Florida (FL)** dominate both count and total contributions, reflecting their political and economic influence.
* **New York (NY) and Illinois (IL)**, known for financial hubs and influential donors, also rank highly.

**High-Value Donors and PACs:**
* Individuals like **Timothy Mellon** ($55 million) and **Michael Bloomberg** contribute heavily, underscoring the disproportionate influence of wealthy individuals.
* Organizations like **One Nation** and **Securing American Greatness** play a pivotal role in aggregating large funds, emphasizing the power of PACs in elections.

**Occupational and Sectoral Influence:**
* High-value contributors predominantly come from sectors like **Investments and Entrepreneurship**, showcasing the financial industry's role in political funding.

**Top Committees:**
* Committees like **C00825851** received record single-day contributions (e.g., $50 million on 07/15/2024), often correlating with campaign milestones.

**Patterns and Trends:**
* **Dominance of Key States:** Larger, wealthier states contribute significantly to campaign funding.
* **Role of PACs:** Super PACs like One Nation aggregate substantial amounts to influence elections.
* **Influence of Wealthy Individuals** High-profile donors like Timothy Mellon shape campaign financing disproportionately.
* **Geographic Concentration:** Contributions are unevenly distributed, favoring states with economic and political power.

**Insights from FEC Validation:**
* **Timothy Mellon**’s ($50) million aligns with his history of substantial conservative funding.

* Insights validated from https://www.politico.com/news/2024/08/20/timothy-mellon-maga-inc-00175249. According to FEC records, Timothy Mellon made a significant contribution of $50 million to the pro-Trump super PAC, Make America Great Again Inc., in July 2024. This aligns closely with the query result, confirming his substantial financial involvement in political campaigns.

* **One Nation**’s $18.4 million contribution reflects its role as a key political player.

* Insights validated from https://www.fec.gov/data/committee/C00468447. One Nation is a registered political action committee (PAC) with the FEC. While specific contribution amounts may vary over time, One Nation has a history of substantial financial involvement in political campaigns, consistent with the query result.

**Reflections:**
* **Concentration of Wealth:** The dominance of wealthy individuals and PACs raises concerns about voter equity in campaign influence.
* **Transparency Needs:** The scale of contributions underscores the importance of campaign finance laws and the need for reforms to ensure fairness.

**Final Conclusions/Key Takeaways:**

* The outcomes of the query align closely with expectations based on historical trends in campaign financing.
* States like California, Texas, and Florida consistently dominate in contributions due to their large populations, economic strength, and political significance
* High-value donors such as Timothy Mellon and PACs like One Nation reflect the expected influence of wealthy individuals and organizations.
* Scale of contributions, including Mellon's $55 million and the $50 million single-day donation to C00825851, underscores the growing reliance on ultra-wealthy donors and PACs.
* This analysis serves as a reminder of the importance of transparency and regulation in maintaining a fair democratic process.